File: src/Java/adventofcode2017/Day11.java

package adventofcode2017;

import utils.Input;

/**
 * https://adventofcode.com/2017/day/11
 */
public class Day11 {

    public static void main(String[] args) throws Exception {
        String input = new Input(2017, 11).strings().get(0);

        int x = 0, y = 0;
        int max = 0, steps = 0;
        for (String s : input.split(",")) {
            switch(s) {
                case "nw" : y -= 1; x -= 1; break;
                case "n" : y -= 1; break;
                case "ne" : x += 1; break;
                case "sw" : x -= 1; break;
                case "s" : y += 1; break;
                case "se" : y += 1; x += 1; break;
            }
            steps = (x > 0 ^ y > 0)
                ? steps = Math.abs(x - y)
                : Math.max(Math.abs(x), Math.abs(y));
            max = Math.max(max, steps);
        }
        System.out.println(steps);
        System.out.println(max);
    }
}

File: src/Python/2019/Day11.py

import math
from utils import Input, printResult
from intcode import Intcode

# https://adventofcode.com/2019/day/11

input = Input(2019, 11).lines()[0]

def runWith(code):
    queue = [code]
    if code == 1:
        queue = [1] + [0] * 1000000
    robot = Intcode(input, queue).run()
    field = {}
    p = (0, 0)
    d = 0
    dirs = {0: [-1, 0], 1: [0, 1], 2: [1, 0], 3: [0, -1]}
    while True:
        color = next(robot, None)
        if color is None:
            break
        field[p] = color
        d = (d - 1 + next(robot) * 2) % 4
        p = (p[0] + dirs[d][0], p[1] + dirs[d][1])
        queue.append(field.get(p, 0))
    return field

field = runWith(0)
printResult(1, len(field))

field = runWith(1)
miny = min(p[0] for p in field)
maxy = max(p[0] for p in field)
minx = min(p[1] for p in field)
maxx = max(p[1] for p in field)

id = ""
palette = {0: "  ", 1: "@@"}
for y in range(miny, maxy+1):
    id  += "\n" + "".join(
        palette[field.get((y, x), 0)]
        for x in range(minx, maxx+1))

printResult(2, id)
